### PR TITLE
[SYCL][L0] Add new environment variable to control the use of copy engine for in-order queues

### PR DESCRIPTION
**Pull request by asudarsa**

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -60,6 +60,16 @@ static const bool UseCopyEngineForD2DCopy = [] {
   return (CopyEngineForD2DCopy && (std::stoi(CopyEngineForD2DCopy) != 0));
 }();
 
+// This is an experiemental option that allows the use of copy engine, if
+// available in the device, in Level Zero plugin for copy operations submitted
+// to an in-order queue. The default is 0.
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:65]: Is it intentionally disabled by default?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -60,6 +60,16 @@ static const bool UseCopyEngineForD2DCopy = [] {
...
+static const bool UseCopyEngineForInOrderQueue = [] {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:66]: I do not see any uses of it, why?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (xtian-github) [reply sycl/plugins/level_zero/pi_level_zero.cpp:65]: -  bool UseCopyEngine = PreferCopyEngine && Queue->Device->hasCopyEngine();
+  bool UseCopyEngine = (!(Queue->isInOrderQueue()) ||
+                        UseCopyEngineForInOrderQueue) && PreferCopyEngine &&
+                        Queue->Device->hasCopyEngine();
 
@smaslov-intel  @asudarsa   Arvind, did you miss above code in the merge?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (xtian-github) [reply sycl/plugins/level_zero/pi_level_zero.cpp:66]: @asudarsa  @smaslov-intel  Use is below. Arvind, make sure you did not miss it.
Old
===
  bool UseCopyEngine = PreferCopyEngine && Queue->Device->hasCopyEngine();

new 
==== 
  bool UseCopyEngine = (!(Queue->isInOrderQueue()) ||
                           UseCopyEngineForInOrderQueue) && PreferCopyEngine &&
                          Queue->Device->hasCopyEngine();

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:65]: Hi @xtian-github
Can you please comment if we want to keep this option turned 'on' or 'off' by default?
And, yes, it does look like I missed these lines during the patch creation. I will fix it now. Thanks

Comment (xtian-github) [reply sycl/plugins/level_zero/pi_level_zero.cpp:65]: @asudarsa  for this new experimental ENV, it should be default OFF, as I did in the patch I give you. We will use it for performance tuning and figure perf regression. For the original COPY engine ENV, it should be default ON (no change to it).

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:65]: We aren't adding NEW env var, we are extending existing one. Let's have all copy engines ON by default and let people use env var to scale it back to deal with any possible regressions

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:65]: Maybe we can make it ON by default and have people use env var to deal with regressions?

Comment (xtian-github) [reply sycl/plugins/level_zero/pi_level_zero.cpp:65]: I would be OK with it. My worry is people may be surprised with regressions.  What is the extension for the syntax?

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I have marked it ON by default. If the user does not specify it, copy engine will be used for in-order queue copies.

**Comment by xtian-github** [comment]
please cherry-pick to xmain. Thanks.

**Review by smaslov-intel — DISMISSED** [review]

**Comment by bader** [comment]
@asudarsa, please, resolve merge conflicts.

**Review by xtian-github — DISMISSED** [review]

**Review by smaslov-intel — DISMISSED** [review]

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**